pkg/kube/informer.go

```diff
@@ -0,0 +1,295 @@
+package kube
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+type InformerFactory interface {
+	Start(stopCh <-chan struct{})
+	ForResource(resource schema.GroupVersionResource) (InformerLister, error)
+	WaitForCacheSync(stopCh <-chan struct{})
+}
+
+type Informer interface {
+	AddEventHandler(handler cache.ResourceEventHandler)
+	AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration)
+	AddIndexers(indexers cache.Indexers) error
+	HasSynced() bool
+}
+
+type Lister interface {
+	List(out runtime.Object, opts ...ListOption) error
+	Get(key types.NamespacedName, out runtime.Object) error
+}
+
+type InformerLister interface {
+	Informer
+	Lister
+}
+
+type multiNamespaceInformerFactory struct {
+	namespaceToFactory map[string]informers.SharedInformerFactory
+}
+
+var _ InformerFactory = &multiNamespaceInformerFactory{}
+
+type NewInformerFactoryFunc func(namespace string) informers.SharedInformerFactory
+
+func NewInformerFactory(namespaces []string, f NewInformerFactoryFunc) (InformerFactory, error) {
+	// TOOD(bison): Dedupe namespaces and check for metav1.NamespaceAll.
+	if len(namespaces) < 1 {
+		return nil, errors.New("must provide at least one namespace, which may be metav1.NamespaceAll")
+	}
+
+	factory := &multiNamespaceInformerFactory{
+		namespaceToFactory: make(map[string]informers.SharedInformerFactory),
+	}
+
+	for _, ns := range namespaces {
+		factory.namespaceToFactory[ns] = f(ns)
+	}
+
+	return factory, nil
+}
+
+func (f *multiNamespaceInformerFactory) ForResource(resource schema.GroupVersionResource) (InformerLister, error) {
+	informer := &multiNamespaceInformer{
+		namespaceToInformer: make(map[string]cache.SharedIndexInformer),
+		resource:            resource,
+	}
+
+	for ns, factory := range f.namespaceToFactory {
+		genericInformer, err := factory.ForResource(resource)
+		if err != nil {
+			return nil, err
+		}
+
+		informer.namespaceToInformer[ns] = genericInformer.Informer()
+	}
+
+	return informer, nil
+}
+
+func (f *multiNamespaceInformerFactory) Start(stopCh <-chan struct{}) {
+	for _, factory := range f.namespaceToFactory {
+		factory.Start(stopCh)
+	}
+}
+
+func (f *multiNamespaceInformerFactory) WaitForCacheSync(stopCh <-chan struct{}) {
+	var wg sync.WaitGroup
+
+	for i := range f.namespaceToFactory {
+		factory := f.namespaceToFactory[i]
+		wg.Add(1)
+		go func() {
```

Comment:
Do we need this to be parallel? My understanding was WaitForCacheSync is just basically polling a atomic variable so it should be the same speed parallel or serial?

Reply:
Yeah, probably not. I think it was just an knee-jerk reaction to calling something with "wait" in the name. Happy to simplify this.